keps/sig-apps/3973-consider-terminating-pods-deployment/README.md

```diff
@@ -0,0 +1,752 @@
+  new pods immediately.
+- Scaling up a Deployment also does not wait for terminating pods and creates (schedules) new pods
+  right away.
+
+Unfortunately, in Deployments with a Recreate strategy we can get mixed behavior. The
+deployment will wait for old pods to terminate during a rollout, but will ignore the terminating
+pods when scaling the pods. So it is still possible to end up with a larger number of pods than
+`.spec.replicas`.
+
+### User Stories (Optional)
+
+#### Story 1
+
+As an application user, I would prefer predictable number of pods in my cluster to prevent any
+scheduling issues and unnecessary autoscaling of nodes. I would also like to achieve consistent
+allocation of other scarce resources to pods.
+
+#### Story 2
+
+As an application user, I would like to keep the old behavior of fast scaling of pods and do not
+mind the higher utilization of resources.
+
+#### Story 3
+
+As an application user, I would like to track the number of instances that perform useful work
+during the entire lifecycle of a pod.
+
+### Notes/Constraints/Caveats (Optional)
+
+### Risks and Mitigations
+
+## Design Details
+
+### Deployment Behavior Changes
+
+Recreate rollout logic:
+- Terminating:
+    1. Scale down old ReplicaSet(s) to 0.
+    2. Wait until all the pods are at least terminating.
+    3. Create new replica set.
+- Terminated: Current behaviour.
+
+RollingUpdate rollout logic:
+- Terminating: Current behaviour.
+- Terminated: When checking if a new replica set can be scaled up, we should consider terminating
+  pods of all ReplicaSets as well and not spawn an amount of replicas that would be higher than
+  `spec.replicas + .spec.strategy.rollingUpdate.maxSurge`.
+
+Scaling logic:
+- Terminating: Current behaviour.
+- Terminated: 
+    - When scaling up across one or more ReplicaSets, we should consider terminating pods of all
+        ReplicaSets as well and not spawn replicas that would be higher than
+        `spec.replicas + .spec.strategy.rollingUpdate.maxSurge`.
+    - Changing scaling down logic is not necessary, and we can scale down as many pods as we want
+        because the policy does not affect this since we are not replacing the pods.
+
+### ReplicaSet Status and Deployment Status Changes
+
+We should keep the current counting behavior for `.status.replicas` regardless of the
+PodReplacementPolicy, for backwards compatibility reasons. Current consumers of the Deployment API
+are only expecting non-terminating pods to be present in this field.
+
+To satisfy the requirement for tracking terminating pods, and for implementation purposes,
+we propose a new field `.status.terminatingReplicas` to the ReplicaSet's and Deployment's
+status.
+
+### API
+
+```golang
+// DeploymentPodReplacementPolicy specifies the policy for creating Deployment Pod replacements.
+// Default is a mixed behavior depending on the DeploymentStrategy
+// +enum 
+type DeploymentPodReplacementPolicy string
```

Comment:
IMO I would maybe ask in api-review to see if we can get some kind of consensus on name.

Reply:
I only saw Terminal being used in reference to the phases. I think Terminated is less ambiguous. 

But for sure, let's bring this up in the api review.

Reply:
Maybe another option is to explicitly name this SucceededOrFailed? Was this already discussed in api review? Job has TerminatingOrFailed in their case, right?

I actually like Terminated, just thought about bringing this up

Reply:
the PR still has to go through API review

- IIRC for Jobs, failed has a different meaning than just the pod status. The pod may fail or may not according to the `PodFailurePolicy` or if it started terminating prematurely (has deletionTimestamp). So it does not always correlate with the pod status. 
- jobs are also using different combinations: TerminatingOrFailed, which is not applicable for Deployments

Since deployments do not care about the final phase of the pod, it might be simpler to call this Terminated. And to have less confusion when comparing this API to the Jobs API.

Reply:
I am not sure if we want to introduce a new terminology (`final`), but rather have some term that can be associated with the pod lifecycle. 

I agree, we can revisit this discussion in the implementation phase.

Reply:
Naming idea:

`podReplacementPolicy: TerminationStarted`
`podReplacementPolicy: TerminationComplete`

(I agree this does not need to be decided until implementation)

Reply:
+1 for `TerminationStarted`/`TerminationComplete`. Still, it is a bit different from the Jobs API, but I suppose if we are already incompatible, we can do the naming a bit differently.

Reply:
I have updated the KEP to the suggestion above.